bot/utils/persistent_scheduling.py

```diff
@@ -0,0 +1,161 @@
+        tasks = await self.cache.items()
+        for task_id, timestamp in tasks:
+            time = Arrow.utcfromtimestamp(timestamp)
+            self._scheduler.schedule_at(time.datetime, task_id, coroutine=self._to_schedule(task_id))
+
+    async def delete(self, task_id: CacheKey) -> None:
+        """
+        Unschedule the task identified by `task_id` and delete it from the cache.
+
+        `cancel` is left as-is in order to be able to pause execution until the next reschedule.
+        Log a warning if the task doesn't exist.
+        """
+        await self.cache.delete(task_id)
+
+        self._scheduler.cancel(task_id)
+
+    async def delete_all(self) -> None:
+        """
+        Unschedule all known tasks and wipe the cache.
+
+        `cancel_all` remains untouched so that it can be used in instances such as cog unloads while preserving cache.
+        """
+        self.reschedule_task.cancel()
+        await self.cache.clear()
+
+        self.reschedule_task.add_done_callback(self._scheduler.cancel_all())
+
+    async def schedule_at(self, time: datetime, task_id: CacheKey) -> None:
+        """
+        Schedule `coroutine` to be executed at the given `time` and add it to the cache.
+
+        If `time` is timezone aware, then use that timezone to calculate now() when subtracting.
+        If `time` is naïve, then use UTC.
+
+        If `time` is in the past, schedule `coroutine` immediately.
+
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
+        """
+        self._scheduler.schedule_at(time, task_id, self._to_schedule(task_id))
+
+        if time.tzinfo is None:
+            time.replace(tzinfo=timezone.utc)
+        await self.cache.set(task_id, time.timestamp())
+
+    async def schedule_later(self, delay: t.Union[int, float], task_id: CacheKey) -> None:
+        """
+        Schedule `coroutine` to be executed after the given `delay` number of seconds and add it to the cache.
+
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
+        """
+        self._scheduler.schedule_later(delay, task_id, self._to_schedule(task_id))
+
+        time = arrow.utcnow() + timedelta(seconds=delay)
+        await self.cache.set(task_id, time.timestamp())
+
+    def cancel_all(self) -> None:
+        """Unschedule all known tasks."""
+        self.reschedule_task.cancel()
+        self.reschedule_task.add_done_callback(lambda _: self._scheduler.cancel_all())
+
+    async def _start_scheduler(self) -> None:
+        """Starts the persistent scheduler by awaiting the guild before rescheduling the tasks."""
+        await self.bot.wait_until_guild_available()
```

Comment:
In what format? I can avoid keeping the bot instance as an attribute, and call the rescheduling separately from outside, but it doesn't seem to be better in terms of coupling. Otherwise, I need the rescheduling task object for the cancels and deletes to work correctly, as you mentioned above.

Reply:
> I can avoid keeping the bot instance as an attribute, and call the rescheduling separately from outside, but it doesn't seem to be better in terms of coupling.

Yes, that is exactly what I mean. The rescheduling function is part of the public interface anyway. It is better for coupling since it won't rely on `Bot` or even an event loop (for `create_task`).

Reply:
Just to be sure, you mean you want the rescheduling function to receive a bot instance and create the task there?
Or do you mean you want whatever instantiated the class to create the task? because then I still need an indicator inside the scheduler that the rescheduling finished.

Reply:
Originally, I meant that the user of the scheduler would call `reschedule_all_tasks` and keep track of the task itself, if needed. The responsibility of ensuring rescheduling is finished before attempting to delete something would be delegated to the user of the class.

However, on second thought, having to manage the task externally sounds inconvenient. I'd feel fine with passing an event loop to the scheduler so it can schedule task. As for waiting for the guild to be available, that could be delegated to the task factory given; not all tasks will necessarily rely on the guild anyway.

Reply:
An alternative is to bring back the event object, and then the task can be created from outside.

Reply:
Well, what do you think about just passing an event loop?

Reply:
Yeah, that's fine, just throwing an idea